### PR TITLE
Fix the server hang in case of long prompts/requests in certain serve…

### PR DESCRIPTION
**Pull request by elvin-n**
…r configuration

There is a discrepancy between limitations of the model, user and cache. To avoid infinite loop we need to discard long requests exceeding server limitations and/or handle long requests more agile

The improper situation can happen:
1. due to small amount of memory on card and required max_num_batched_tokens leading to small amount of the kv cache
2. if we cancel some requests it might happen that request removes from active batch and then when we can add it again, we might have situatino when number of decoded tokens is bigger than max_num_batched_tokens that also lead to server hang

@sunggg @masahi @binarybana

### CODE REVIEW

**Comment by sunggg** [comment]
Thank you @elvin-n for the improvement! I like how this PR aims to drop the hyperparameter. Would you share how I can reproduce this hanging issue? And I'm wondering what is the relation with our current approach landed by PR https://github.com/octoml/mlc-llm/pull/85 which we error out at the initialization time when the kv cache blocks are not sufficient.

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 1]
```diff
@@ -194,24 +197,25 @@ def step(self) -> GenerationLoopWorkerOutput:
 
         self.stopped_requests.clear()
 
-        for state in self.cancelled_requests:
-            err = None
-            if state.validation_err:
-                err = state.validation_err
+        with self.queue_lock:
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py:199]: Why do we need this lock?

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 2]
```diff
@@ -114,8 +112,13 @@ def add(self, request_states: list[RequestState]):
                 if (
                     request_state.validation_err is not None
                     or request_state.prompt_len >= self.max_context_length
+                    or request_state.prompt_len >= self.max_num_batched_tokens > 0
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py]: Let's clean up the two conditions as 
`request_state.prompt_len >= min(self.max_context_length, self.max_num_batched_tokens)`

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 3]
```diff
@@ -114,8 +112,13 @@ def add(self, request_states: list[RequestState]):
...
+                    or (self.cache_manager.get_kv_cache_size() - request_state.prompt_len) < 0
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py]: Why do we need this? At this point we know that `prompt_len < max_num_batched_tokens`. So we can process this request, after evicting the cache if necessary.

Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py]: Why do we need this? At this point we know that `prompt_len < max_num_batched_tokens`. So we can process this request, after evicting the cache if necessary.

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 4]
```diff
@@ -114,8 +112,13 @@ def add(self, request_states: list[RequestState]):
...
+                    or (self.cache_manager.get_kv_cache_size() - request_state.prompt_len) / (len(self.current_batch) + 1) < self.max_decode_steps
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py]: Again, assuming that the cache eviction is working, I don't understand why we need this condition.

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 5]
```diff
@@ -293,15 +297,21 @@ def _adjust_batch(self):
                 state = self.queue[0]
                 num_tokens = len(state.token_ids)
                 num_new_batched_tokens += num_tokens
+                # this can happen if we processed some request and then had to remove it from inference
+                # due to limit of the kv cache. It can appear that new size to process is bigger
+                # than max_num_batched_tokens. Need to roll back to acceptable size
+                if not len(self.current_batch) and num_new_batched_tokens > self.max_num_batched_tokens:
+                    state.token_ids = state.token_ids[:self.max_num_batched_tokens]
+                    state.next_start_position = num_new_batched_tokens = num_tokens = self.max_num_batched_tokens
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py:310]: I don't understand this change. Please add an example or test case.

File: serve/mlc_serve/engine/base.py [hunk 6]
```diff
@@ -27,7 +27,6 @@ class MLCServeEngineConfig:
     max_num_batched_tokens: int = -1
     min_decode_steps: int = 32
     max_decode_steps: int = 48
-    prompt_allocate_ratio: float = 2.0
```
Comment (masahi) [inline serve/mlc_serve/engine/base.py:30]: Why remove this? I remember this is one of perf tuning knobs. So it is not supposed to be removed without a good reason.

**Comment by elvin-n** [comment]
> Thank you @elvin-n for the improvement! I like how this PR aims to drop the hyperparameter. Would you share how I can reproduce this hanging issue? And I'm wondering what is the relation with our current approach landed by PR #85 which we error out at the initialization time when the kv cache blocks are not sufficient.

We have quite complex logic in `_adjust_batch` which depend on cache size, max batch size, some assumption if there is a room for new request. By fact, we had condition for max_decode_steps, min_decode_step and prompt_allocate_ratio. Each of them could be broken by next requests. For example, if we decide that request can be added into the current batch because it fit to cache with `prompt_allocate_ratio` scale, then next request also can be added into the batch and invalidate decision for the first request. But this previous request is not removed from current batch. No sense to have a condition with `prompt_allocate_ratio` if it can be failed in the same loop on the next iteration.

Due this reason I left only comparison to the `max_decode_steps` that should be treated as "Add a request if ther eis some room for decode". This space is not big enough. We rather stop to add by this condition rather than stop adding due to `prompt_allocate_ratio`. BTW, I am not sure why we need to have `max_decode_steps` and `min_decode_steps`. We are adding new requests if we satisfy min, but stop adding if not satisfy `max_decode_steps`. Looks slightly incosistent.

Regarding how to reproduce the issue - I ran server with `--max-num-sequences 10 --max-input-len 20` and forcedly limited in [this place](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/model/paged_cache_model.py#L745) number of blocks by 128. Then run our benchmark_servin.py for 100 or 1000 requests.

File: serve/mlc_serve/engine/base.py (hunk 6 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/base.py:30]: answer in this https://github.com/octoml/mlc-llm/pull/89#issuecomment-1828641261 comment.

**Comment by masahi** [comment]
> Regarding how to reproduce the issue - I ran server with --max-num-sequences 10 --max-input-len 20 and forcedly limited in [this place](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/model/paged_cache_model.py#L745) number of blocks by 128

To avoid confusion / misuse, I think we should require `max-batched-tokens` (or equivalent) to be passed at command line. That `num_blocks = 500` is a historical artifact, and 500 or 128 cache blocks is too few.

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 3 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: assuming that PR #85 verifies this and get_kv_cache_size > max_num_batched_tokens your comment is correct, we can remove this condition

**Comment by elvin-n** [comment]
> > Regarding how to reproduce the issue - I ran server with --max-num-sequences 10 --max-input-len 20 and forcedly limited in [this place](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/model/paged_cache_model.py#L745) number of blocks by 128
> 
> To avoid confusion / misuse, I think we should require `max-batched-tokens` (or equivalent) to be passed at command line. That `num_blocks = 500` is a historical artifact, and 500 or 128 cache blocks is too few.

Why num_blocks cannot be 128 or less? it depends on the topology, amount of memory allocated by inference function self.mod["evaluate"] during memory profiling and amount of memory on GPU.
We can say that "we want to have it big enough" but real values of cahce size can be very different even during regular start of the server without my hacks

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 3 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: we do not need the first condition (will remove), but we still need the second one. It is a verification that this request ill have a chance to appear in the current_batch. It relate to the condition with `prompt_allocate_ratio` that I substituted in `_adjust_batch`. If we do not have such verification in this place, we can have a situation when we will not be able to add such reqquest to current batch and this one of the use case when server hang could appear

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 4 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: for example, previously we had `prompt_allocate_ratio == 1.5`, `max_num_batched_tokens == 200`, `cache size == 250` and you get a request with prompt == 200. It passes all your verification, request is added into list to be processed, but in the `_adjust_batch` you had a verification that prompt_len*prompt_allocate_ratio not less than cache size. But for above parameters, such verification failed and we received infinite attempts to add and each time this attempt fails. This is a reason of the hang. 

I substituted verification in the _adjust_batch vs `prompt_allocate_ratio` by verification of the enough room for each request - `self.cache_manager.get_max_new_tokens() <= self.max_decode_steps` but I still need to verify if we have an ability to add a request for inference and get rid of it on the submissin stage to avoid infinit attempts in `_adjust_batch`. This is a reason of the second condition in this place.

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 1 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py:199]: I got an impression that I observed some data race during debugging when we had requests which should not be in the `self.cancelled_requests`. Probably because we add requests into self.canceled_requests in [add()](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/engine/staging_engine_worker.py#L118C24-L118C24) and remove in [step()](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/engine/staging_engine_worker.py#L214) that I assume can be called concurrently from different threads. I guarded place in `step()`, place in `add()` was guarded before

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 5 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py:310]: you have two requests, each of the prompt_len == 50, cache_size == 150 and max_num_batched_tokens == 70. Each are decoded until 100 new tokens. During the decode we decoded 25 tokens for each request, `_add_batch` understands that it requires more space to continue processing at least one of the requests, and remove one of the request from processing [here](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/engine/staging_engine_worker.py#L269C29-L269C29). When we finish processing of the first, we would like to take the second. The problem that it will have a number of tokens equal to 75 because it is not only prompt but [prompt + previously decoded tokens](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/engine/staging_engine_worker.py#L294C36-L294C36) and this len exceeds our limit of `max_num_batched_tokens` and we again got an infinite loop

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 2 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: I am confused that we have a comparison of `self.max_num_batched_tokens > 0`. It means that at the moment when we added such condition, we were not sure if max_num_batched_tokens can be zero. Are we sure that `max_num_batched_tokens` is always positive?

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 3 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: fixed

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 2 above)
Comment (masahi) [reply serve/mlc_serve/engine/staging_engine_worker.py]: `max_num_batched_tokens` is of course always positive. I thought `> 0` was a typo.

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 1 above)
Comment (masahi) [reply serve/mlc_serve/engine/staging_engine_worker.py:199]: yeah that makes sense.

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 2 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: fixed

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 4 above)
Comment (masahi) [reply serve/mlc_serve/engine/staging_engine_worker.py]: ok, I understand that the cache size must be bigger than `prompt_allocate_ratio * max_num_batched_tokens` to prevent a hang. But I still don't understand why the upper bound is `max_decode_steps`.

Comment (masahi) [reply serve/mlc_serve/engine/staging_engine_worker.py]: Also I realized that both the existing and the new conditions are suboptimal for Mistral since with SWA not all prompt tokens need to have its KV values cached. We implemented such optimization in https://github.com/mlc-ai/mlc-llm/blob/main/mlc_llm/relax_model/llama_batched_vllm.py#L89-L92

So for example, 

```
if (
    self.cache_manager.get_free_space()
    <= self.prompt_allocate_ratio * num_tokens
):
```
can be optimized to

```
if (
    self.cache_manager.get_free_space()
    <= self.prompt_allocate_ratio * min(num_tokens, window_size)
):
```

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 7]
```diff
@@ -293,15 +295,21 @@ def _adjust_batch(self):
...
                     LOG.debug(
                         "Stop growing the batch due to max_num_batched_tokens. Batched tokens: %s",
                         num_new_batched_tokens,
                     )
                     break
                 if (
-                    self.cache_manager.get_free_space()
-                    <= self.prompt_allocate_ratio * num_tokens
+                    (self.cache_manager.get_free_space() - num_tokens) < 0 or \
+                         (self.cache_manager.get_free_space() - num_tokens) / (len(self.current_batch) + 1) < self.max_decode_steps
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py]: Seems `(self.cache_manager.get_free_space() - num_tokens) < 0` is redundant? It is subsumed by the second condition.

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 4 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: I do not see reason for leaving of `prompt_allocate_ratio`, as I pointed it can be easily broken by the next request on the next iteration of the loop. So, if we have situation which breaks this condition, it's no sense to have it.

As for sliding window - seems yes, it should be improved in two places 1) for requests which is preempted from kv cache should be cut if len(prompt) exceed `max_num_batched_tokens` but not from begin, but from the end of token list, right? 2) in case of new comparison instead of `prompt_allocate_ratio`, we need to take into account `window_size` as well.

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 5 above)
Comment (masahi) [reply serve/mlc_serve/engine/staging_engine_worker.py:310]: Ok, so after recovering from a cache eviction, we might need to cache both prompt and intermediate decode tokens, sum of which could be `> max_num_batched_tokens`.

Comment (masahi) [reply serve/mlc_serve/engine/staging_engine_worker.py:310]: Here we can also exploit SWA, since we only need to "recache" `window_size + len(decode_tokens)` KV tensors.

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 4 above)
Comment (masahi) [reply serve/mlc_serve/engine/staging_engine_worker.py]: > 1) for requests which is preempted from kv cache should be cut if len(prompt) exceed max_num_batched_tokens but not from begin, but from the end of token list, right?

Your sentence is not making a sense, but I guess yes, we only need to cache the most recent token `window_size` tokens.

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 7 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: yep, you are right

Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: fixed

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 5 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py:310]: > we might need to cache both prompt and intermediate decode tokens, sum of which could be > max_num_batched_tokens.

We cannot do caching until we do a prefill which is limited by `max_num_batched_tokens`. Exactly prefill will create key-values that will be cached. Until we do prefill, there is no something that can be put into kv_cache

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 2 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: > max_num_batched_tokens is of course always positive. I thought > 0 was a typo.

It appeared not true. At least the unit test assumed that `max_num_batched_tokens` can be uninitialized and was equal to -1

Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: and due this reason we [have comparison](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/engine/sync_engine.py#L276) on positive value in current `_adjust_batch`

@yelite @sunggg We need to decide if we can have `max_num_batched_tokens` always defined or not

Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: I can fix test and add assert into StagingInferenceEngine/SynchronousInferenceEngine initialization. Just need to understand if we see any inference Model that is not limited by maximum number of tokens

Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: ok, it seems it was something initially for testing, but it cannot work since logic of engines depends on certain assumptions. Eventually I removed ability not to check configuration in [base.py:42](https://github.com/octoml/mlc-llm/blob/batch-serving/serve/mlc_serve/engine/base.py#L42) and modified test to pass verification. All tests passed

**Comment by elvin-n** [comment]
Probably we need to spend a day on attempt to create unit test for corner cases fixed in this PR

**Comment by masahi** [comment]
> Probably we need to spend a day on attempt to create unit test for corner cases fixed in this PR

@elvin-n If you can create a test case that demonstrates the hang, I can also use that to implement and test "hang detection".

**Comment by elvin-n** [comment]
> > Probably we need to spend a day on attempt to create unit test for corner cases fixed in this PR
> 
> @elvin-n If you can create a test case that demonstrates the hang, I can also use that to implement and test "hang detection".

Yes, will do with specially developed DummyModelModule which will emulate behaviour of kv_cache in PagedCacheModelModule and leading to the hang in both engines

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 2 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: had to rollback changes back because for context we have not less, but for `max_num_batched_tokens` we have to exclude only in case of exceeding. Boarding conditions are processed wrong if we have only `min(...)`

File: serve/tests/unittest/test_staged_inference_engine.py [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+        if not len(s.outputs):
+            empty_step += 1
+
+    assert len(finished) == 2
+    assert empty_step < 10
+
+
+# Test to verify if new comming request with big prompt can be put into inference
+# and does not have issues with cache size limits verification
+def test_big_prompt_fit_to_cache_staging():
```
Comment (masahi) [inline serve/tests/unittest/test_staged_inference_engine.py]: This test doesn't hang even without your engine change. Is that expected?

File: serve/tests/unittest/test_staged_inference_engine.py [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+        if stage == 4 and len(s.outputs) > 1:
+            stage = 5
+
+    assert stage == 4
+
+
+# Test to verify if evicted request from active batch which in intermediate
+# state exceeding the max_num_batched_tokens can be processed successfully and will
+# not hang the server in infinite attempt to return it back to the active loop
+def test_cache_evict_hang_staging():
```
Comment (masahi) [inline serve/tests/unittest/test_staged_inference_engine.py]: I confirmed that this test hangs without the engine changes in this PR, and I get the hang warning you added in https://github.com/octoml/mlc-llm/pull/91 as expected. Great! cc @jroesch @itay

File: serve/tests/unittest/test_staged_inference_engine.py (hunk 8 above)
Comment (masahi) [reply serve/tests/unittest/test_staged_inference_engine.py]: ok using `prompt_allocate_ratio = 2.0` instead of 1.0 in `DummaryModelModule`, I get a hang from this test as well.

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 4 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py]: WSA will be addressed in separate request

File: serve/tests/unittest/test_staged_inference_engine.py [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+    DecodeRequest,
+    KVCache,
+    PrefillRequest,
+    SequenceId,
+    TextGenerationResult,
+)
+from mlc_serve.engine.sync_engine import SynchronousInferenceEngine
+from mlc_serve.engine.staging_engine import StagingInferenceEngine
+
+from tests.unittest.dummary_model import (
```
Comment (masahi) [inline serve/tests/unittest/test_staged_inference_engine.py]: Please make it possible to run the test from the `mlc-llm` root directory.

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 11]
```diff
@@ -114,8 +112,12 @@ def add(self, request_states: list[RequestState]):
                 if (
                     request_state.validation_err is not None
                     or request_state.prompt_len >= self.max_context_length
+                    or request_state.prompt_len > self.max_num_batched_tokens
+                    or self.cache_manager.get_kv_cache_size() - request_state.prompt_len < self.max_decode_steps
                 ):
                     self.cancelled_requests.append(request_state)
+                    if request_state.validation_err is None:
+                        request_state.validation_err = "Server configuration unable to process the request of such prompt length."
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py]: "The prompt is too long for the given set of engine parameters."

File: serve/mlc_serve/engine/sync_engine.py [hunk 12]
```diff
@@ -98,8 +97,15 @@ def add(self, requests: list[Request]):
...
-            if state.prompt_len >= self.max_context_length:
+            if (
+                state.validation_err is not None
+                or state.prompt_len >= self.max_context_length
+                or state.prompt_len > self.max_num_batched_tokens
+                or self.cache_manager.get_kv_cache_size() - state.prompt_len < self.max_decode_steps
+            ):
                 self.cancel(req.request_id)
+                if state.validation_err is None:
+                    state.validation_err = "Server configuration unable to process the request of such prompt length."
```
Comment (masahi) [inline serve/mlc_serve/engine/sync_engine.py]: "The prompt is too long for the given set of engine parameters."

File: serve/mlc_serve/engine/sync_engine.py [hunk 13]
```diff
@@ -60,15 +60,14 @@ def __init__(
         assert self.model_artifact_config.max_context_length, "max_context_length must not be zero"
         self.max_context_length = self.model_artifact_config.max_context_length
         self.max_num_batched_tokens = model_module.engine_config.max_num_batched_tokens
+        assert self.max_num_batched_tokens > 0, "max_num_batched_tokens must be defined and be greatly 0"
```
Comment (masahi) [inline serve/mlc_serve/engine/sync_engine.py]: "max_num_batched_tokens must be positive."

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 14]
```diff
@@ -293,15 +296,21 @@ def _adjust_batch(self):
                 state = self.queue[0]
                 num_tokens = len(state.token_ids)
                 num_new_batched_tokens += num_tokens
+                # this can happen if we processed some request and then had to remove it from inference
+                # due to limit of the kv cache. It can appear that new size to process is bigger
+                # than max_num_batched_tokens. Need to roll back to acceptable size
+                if not len(self.current_batch) and num_new_batched_tokens > self.max_num_batched_tokens:
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py]: `len(self.current_batch) == 0`?

**Review by masahi — CHANGES_REQUESTED** [review]
More review coming

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 15]
```diff
@@ -114,8 +112,12 @@ def add(self, request_states: list[RequestState]):
                 if (
                     request_state.validation_err is not None
                     or request_state.prompt_len >= self.max_context_length
+                    or request_state.prompt_len > self.max_num_batched_tokens
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py]: Why do we need `>` here? I don't understand your comment in https://github.com/octoml/mlc-llm/pull/89#discussion_r1409071178. Please document your reasoning in the code as well.

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 16]
```diff
@@ -114,8 +112,12 @@ def add(self, request_states: list[RequestState]):
...
+                    or self.cache_manager.get_kv_cache_size() - request_state.prompt_len < self.max_decode_steps
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py:116]: Document why we need this condition. My understanding is that if we remove this condition and allow such request, the request can never put into the batch due to the condition https://github.com/octoml/mlc-llm/blob/04a27a0f09e7a43608f22e1c99089da390452ddc/serve/mlc_serve/engine/staging_engine_worker.py#L279-L284

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 17]
```diff
@@ -293,15 +296,21 @@ def _adjust_batch(self):
...
                     LOG.debug(
                         "Stop growing the batch due to max_num_batched_tokens. Batched tokens: %s",
                         num_new_batched_tokens,
                     )
                     break
                 if (
-                    self.cache_manager.get_free_space()
-                    <= self.prompt_allocate_ratio * num_tokens
+                    (self.cache_manager.get_free_space() - num_tokens) / (len(self.current_batch) + 1)
+                        < self.max_decode_steps
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py:320]: Document what this condition is for.

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 18]
```diff
@@ -293,15 +296,21 @@ def _adjust_batch(self):
                 state = self.queue[0]
                 num_tokens = len(state.token_ids)
                 num_new_batched_tokens += num_tokens
+                # this can happen if we processed some request and then had to remove it from inference
+                # due to limit of the kv cache. It can appear that new size to process is bigger
+                # than max_num_batched_tokens. Need to roll back to acceptable size
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py]: Here is my proposal for the revised text based on my understanding:

"""
This can happen when we are recovering from cache eviction and the sum of prompt and intermediate decode tokens is bigger than the biggest allowable batch size, `self.max_num_batched_tokens`. In such cases, we need to discard the recent decode tokens that cannot fit into a batch, and recompute them after we fill the cache entries for the older tokens. 
"""

File: serve/mlc_serve/engine/sync_engine.py [hunk 19]
```diff
@@ -273,15 +279,21 @@ def _adjust_batch(self):
...
                     logger.debug(
                         "Stop growing the batch due to max_num_batched_tokens. Batched tokens: %s",
                         num_new_batched_tokens,
                     )
                     break
                 if (
-                    self.cache_manager.get_free_space()
-                    <= self.prompt_allocate_ratio * num_tokens
+                    (self.cache_manager.get_free_space() - num_tokens) / (len(self.current_batch) + 1)
+                        < self.max_decode_steps
```
Comment (masahi) [inline serve/mlc_serve/engine/sync_engine.py:305]: For the change in sync engine, please apply my comments on the staging engine to here as well. Or extract common conditions into a utility that can be used by both engines.

File: serve/tests/unittest/test_staged_inference_engine.py [hunk 20]
```diff
@@ -0,0 +1,322 @@
+from typing import Optional, Union
```
Comment (masahi) [inline serve/tests/unittest/test_staged_inference_engine.py:1]: Please rename this file to `test_staging_engine.py`

File: serve/tests/unittest/test_synchronous_inference_engine.py [hunk 21]
```diff
@@ -2,7 +2,6 @@
```
Comment (masahi) [inline serve/tests/unittest/test_synchronous_inference_engine.py:2]: Please rename this file to `test_sync_engine.py`

File: serve/tests/unittest/dummary_model.py [hunk 22]
```diff
@@ -0,0 +1,139 @@
+from typing import Optional, Union
```
Comment (masahi) [inline serve/tests/unittest/dummary_model.py]: Put the contents of this file under `mlc_serve/model/dummy.py`

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 16 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py:116]: exactly, that will lead to hang. To avoid such situation, we added verification on request add stage and cancel it if we understand that we cannot add such request for processing

File: serve/mlc_serve/engine/sync_engine.py (hunk 19 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/sync_engine.py:305]: > Or extract common conditions into a utility that can be used by both engines.

It's too big refactoring, let's plan it as a separate PR. In other case this PR will never landed )

File: serve/mlc_serve/model/dummary_model.py [hunk 23]
```diff
@@ -0,0 +1,138 @@
+from typing import Optional, Union
```
Comment (masahi) [inline serve/mlc_serve/model/dummary_model.py]: Just `dummy.py` is fine (we already have `engine/dummy.py`). `dummary` is not an English word.

File: serve/mlc_serve/engine/staging_engine_worker.py (hunk 16 above)
Comment (elvin-n) [reply serve/mlc_serve/engine/staging_engine_worker.py:116]: added a test, added comment

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 24]
```diff
@@ -111,11 +108,19 @@ def add(self, request_states: list[RequestState]):
...
             for request_state in request_states:
+                # We need to exclude requests having prompt exceeding max_context_length or
+                # max_num_batched_tokens because model will work in non appropriate mode
+                # or because model cannot process number of tokens in one prefill
+                # invocation
+                # As well we need to exclude requests which cannot fit into the kv_cache
                 if (
                     request_state.validation_err is not None
-                    or request_state.prompt_len >= self.max_context_length
+                    or request_state.prompt_len > min(self.max_context_length,self.max_num_batched_tokens)
```
Comment (masahi) [inline serve/mlc_serve/engine/staging_engine_worker.py]: With this condition, I think the reasoning is obvious. So we can omit a lengthy comments in L111-L115

Comment (masahi) [reply serve/mlc_serve/engine/staging_engine_worker.py]: `min(self.max_context_length, self.max_num_batched_tokens)` (missing space)

**Review by masahi — APPROVED** [review]

File: serve/mlc_serve/engine/staging_engine_worker.py [hunk 25]
```diff
@@ -113,9 +110,14 @@ def add(self, request_states: list[RequestState]):
             for request_state in request_states:
                 if (
                     request_state.validation_err is not None
-                    or request_state.prompt_len >= self.max_context_length
+                    or request_state.prompt_len > min(self.max_context_length, self.max_num_batched_tokens)
```
Comment (sunggg) [inline serve/mlc_serve/engine/staging_engine_worker.py:113]: Now that I think about it again, should we just check with `self.max_num_batched_tokens`? In theory, the engine should be able to produce the output although its quality might be terrible.
